main.cpp

```diff
@@ -62,23 +115,39 @@ static void matrix_multiply(Matrix &out, Matrix const &lhs, Matrix const &rhs) {
     out.reshape(nx, ny);
 
     // 这个循环为什么不够高效？如何优化？ 15 分
-#pragma omp parallel for collapse(2)
-    for (int y = 0; y < ny; y++) {
-        for (int x = 0; x < nx; x++) {
-            out(x, y) = 0;  // 有没有必要手动初始化？ 5 分
-            for (int t = 0; t < nt; t++) {
-                out(x, y) += lhs(x, t) * rhs(t, y);
+    // ans: lhs是跳跃访问，rhs是连续访问，out不动，造成无法矢量化。
+
+// #pragma omp parallel for collapse(2)
+//     for (int y = 0; y < ny; y++) {
+//         for (int x = 0; x < nx; x++) {
+//             out(x, y) = 0;  // 有没有必要手动初始化？ 5 分
+//             for (int t = 0; t < nt; t++) {
+//                 out(x, y) += lhs(x, t) * rhs(t, y);
+//             }
+//         }
+//     }
+
+    #pragma omp parallel for collapse(2)
+    for(int j=0; j<ny; j++){
+        for(int i=0; i<nx; i+=32){
+            for(int t=0; t<nt; t++){
+                for(int i_block=i; i_block<i+32; i_block++){
+                    out(i,j) += lhs(i_block, t) *  rhs(t, j);
+                }
             }
         }
     }
+
     TOCK(matrix_multiply);
 }
 
 // 求出 R^T A R
 static void matrix_RtAR(Matrix &RtAR, Matrix const &R, Matrix const &A) {
     TICK(matrix_RtAR);
     // 这两个是临时变量，有什么可以优化的？ 5 分
-    Matrix Rt, RtA;
+    // ans: 改为static变量，预先分配好空间。
+    static Matrix Rt(std::array<std::size_t, 2>{1024, 1024}), RtA(std::array<std::size_t, 2>{1024, 1024});
```

Comment:
```suggestion
    static thread_local Matrix Rt, RtA;
```
我觉得可以一开始为空没问题。thread_local保证如果多个线程访问不会出错。